src/main/java/de/zalando/zmon/dataservice/data/AbstractWorkResultWriter.java

```diff
@@ -0,0 +1,236 @@
+import com.fasterxml.jackson.databind.node.DecimalNode;
+import com.fasterxml.jackson.databind.node.NumericNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.google.common.collect.ImmutableSet;
+import de.zalando.zmon.dataservice.DataServiceMetrics;
+import de.zalando.zmon.dataservice.config.DataServiceConfigProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.StringUtils;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.regex.Pattern;
+
+public abstract class AbstractWorkResultWriter implements WorkResultWriter {
+  private static final Set<String> DEFAULT_ENTITY_TAG_FIELDS =
+      new HashSet<>(
+          ImmutableSet.of(
+              "application_id",
+              "application_version",
+              "stack_name",
+              "stack_version",
+              "application",
+              "version",
+              "account_alias",
+              "cluster_alias",
+              "alias",
+              "namespace"));
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractWorkResultWriter.class);
+  private static final String TIME_SERIES_METRICS_NAME_PREFIX = "zmon.check.id.";
+  private static final String REPLACE_CHAR = "_";
+  private static final Pattern INVALID_TAG_CHARS = Pattern.compile("[?@:=\\[\\]]");
+
+  private final DataServiceConfigProperties config;
+  private final DataServiceMetrics dataServiceMetrics;
+  private Set<String> entityTagFields;
+
+  public AbstractWorkResultWriter(
+      DataServiceConfigProperties config, DataServiceMetrics dataServiceMetrics) {
+    this.config = config;
+    this.dataServiceMetrics = dataServiceMetrics;
+    if (null == config.getKairosdbTagFields() || config.getKairosdbTagFields().size() == 0) {
+      this.entityTagFields = DEFAULT_ENTITY_TAG_FIELDS;
+    } else {
+      this.entityTagFields = new HashSet<>(config.getKairosdbTagFields());
+    }
+  }
+
+  @Override
+  public void write(WriteData writeData) {
+    if (writeData.getWorkerResultOptional().isPresent()) {
+      final WorkerResult workerResult = writeData.getWorkerResultOptional().get();
+      if (workerResult == null || workerResult.results == null || workerResult.results.isEmpty()) {
+        LOG.warn("Received a request with invalid results: {}", workerResult);
+        return;
+      }
+      try {
+        // Translate to generic data model
+        List<GenericMetrics> metrics = new ArrayList<>();
+        for (CheckData cd : workerResult.results) {
+          formatTimeSeriesMetrics(cd, metrics);
+        }
+        store(metrics);
+
+      } catch (Exception ex) {
+        if (config.isLogKairosdbErrors()) {
+          LOG.error("KairosDB write path failed", ex);
+        }
+          dataServiceMetrics.markKairosError();
+      }
+    }
+  }
+
+  protected abstract void store(List<GenericMetrics> metrics);
```

Comment:
It's better to agree on methods arrangement inside the class. The most common one is `public` -> `protected` -> `package private` -> `private`.